### PR TITLE
Add custom contact handling

### PR DESCRIPTION
**Pull request by micahhainline**
Added support for issue #100 

As part of WWT Asynchrony Labs integration of CareKit with video chat and CallKit, we needed to have more flexibility in how the Connect page handled contact items.

The new mechanism is flexible at a number of levels, while still providing ease of construction for the OCKContact. While the changes were not entirely backward compatible, the original initializer was maintained (though deprecated) to avoid creating breakages for most users.

Our original hope was to follow the original direction of the OCKContact and expand the usage of API from the Contacts framework (such as CNPhoneNumber) but this turned out not to be practical. The mechanisms that the Contacts application would use to interact with CallKit-enabled applications (launching from bundle ids, querying Intents) are not available to us, as they are private. This left us needing to use custom URL schemes or custom handling of the button presses, both of which we have made into easy options within these changes. Making a clean break from the Contacts framework seems like the best approach going forward.

Icons were added for a new video option, which will default to using Facetime. The default label for this was added as "video" and we made a best-effort attempt to localize this into the supported languages, but it would be reasonable to check our work there.

### CODE REVIEW

File: CareKit/Connect/OCKConnectDetailViewController.m [hunk 1]
```diff
@@ -32,7 +33,7 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 #import "OCKConnectDetailViewController.h"
 #import "OCKConnectTableViewHeader.h"
 #import "OCKDefines_Private.h"
-
+#import "OCKHelpers.h"
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectDetailViewController.m]: Add extra new line.

File: CareKit/Connect/OCKContact.h [hunk 2]
```diff
@@ -28,10 +29,8 @@
  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
-
-#import <CareKit/CareKit.h>
 #import <ContactsUI/ContactsUI.h>
-
+#import <CareKit/CareKit.h>
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContact.h]: ?

File: CareKit/Connect/OCKContact.m [hunk 3]
```diff
@@ -32,7 +33,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 #import "OCKContact.h"
 #import "OCKHelpers.h"
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContact.m]: Add extra new line.

File: CareKit/Connect/OCKContactInfo.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+ FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.h]: Add extra new line.

File: CareKit/Connect/OCKContactInfo.h [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.h]: Add extra new line.

File: CareKit/Connect/OCKContactInfo.h [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+NS_ASSUME_NONNULL_BEGIN
+
+typedef NS_ENUM(NSInteger, OCKContactInfoType) {
+	OCKContactInfoTypePhone = 0,
+	OCKContactInfoTypeMessage,
+	OCKContactInfoTypeEmail,
+	OCKContactInfoTypeVideo
+};
+
+@interface OCKContactInfo : NSObject <NSSecureCoding, NSCopying>
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.h]: OCK_CLASS_AVAILABLE

File: CareKit/Connect/OCKContactInfo.h [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+#import <UIKit/UIKit.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef NS_ENUM(NSInteger, OCKContactInfoType) {
+	OCKContactInfoTypePhone = 0,
+	OCKContactInfoTypeMessage,
+	OCKContactInfoTypeEmail,
+	OCKContactInfoTypeVideo
+};
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.h]: Provide documentation for enum. Here's an example:

/**
 An enumeration of the types of contacts available.
 */
OCK_ENUM_AVAILABLE
typedef NS_ENUM(NSInteger, OCKContactType) {
    /**
     A care team contact such as a physician or nurse.
     */
    OCKContactTypeCareTeam = 0,
    
    /**
     A personal contact such as a friend or family member.
     */
    OCKContactTypePersonal
};

File: CareKit/Connect/OCKContactInfo.h [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+@interface OCKContactInfo : NSObject <NSSecureCoding, NSCopying>
+
+@property (nonatomic, readonly) OCKContactInfoType type;
+@property (nonatomic, readonly) NSString *displayString;
+@property (nonatomic, readonly) NSURL *actionURL;
+@property (nonatomic, readonly) NSString *label;
+@property (nonatomic, readonly) UIImage *icon;
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.h]: Provide documentation for all properties of a public class.

File: CareKit/Connect/OCKContactInfo.h [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+/**
+ Creates a new contact information item. The label and icon will be set to the default for the given contact type.
+ 
+ @param type The type of contact info, which will define the display icon and the default action if no actionURL is specified.
+ @param displayString The string that will be displayed to the user to represent the contact item.
+ @param actionURL The action to take to initiate the connection. For example "sms:1-314-555-1234", "tel:1-314-555-1234", "facetime://user@example.com"
+ 
+ @return An instance of the contact info.
+ */
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.h]: Adjust documentation formatting to match CareKit documentation.

File: CareKit/Connect/OCKContactInfo.h [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL;
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.h]: Change `- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL` to `- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(nullable NSURL *)actionURL`.

For consistency.

File: CareKit/Connect/OCKContactInfo.h [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+ Creates a new contact information item. The icon will be set to the default for the given contact type.
+ 
+ @param type The type of contact info, which will define the display icon and the default action if no actionURL is specified.
+ @param displayString The string that will be displayed to the user to represent the contact item.
+ @param actionURL The action to take to initiate the connection. For example "sms:1-314-555-1234", "tel:1-314-555-1234", "facetime://user@example.com"
+ @param label The label of the contact info.
+ 
+ @return An instance of the contact info.
+ */
+- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL label:(NSString *)label;
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.h]: Change - (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL to - (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(nullable NSURL *)actionURL.

For consistency.

File: CareKit/Connect/OCKContactInfo.h [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+ 
+ @param type The type of contact info, which will define the display icon and the default action if no actionURL is specified.
+ @param displayString The string that will be displayed to the user to represent the contact item.
+ @param actionURL The action to take to initiate the connection. For example "sms:1-314-555-1234", "tel:1-314-555-1234", "facetime://user@example.com"
+ @param label The label of the contact info.
+ @param icon The icon which represents the contact type. The contact tint color will be applied to this icon, which will be converted if needed to UIImageRenderingModeAlwaysTemplate.
+ 
+ @return An instance of the contact info.
+ */
+- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL label:(NSString *)label icon:(UIImage * _Nullable)icon;
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.h]: Refactor nullability.

File: CareKit/Connect/OCKContactInfo.h [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
++ (OCKContactInfo *)smsContactInfo:(NSString *)smsNumber;
++ (OCKContactInfo *)phoneContactInfo:(NSString *)phoneNumber;
++ (OCKContactInfo *)emailContactInfo:(NSString *)emailAddress;
++ (OCKContactInfo *)facetimeVideoContactInfo:(NSString *)emailAddressOrRawPhoneNumber displayString:(NSString * _Nullable)displayString;
++ (OCKContactInfo *)facetimeAudioContactInfo:(NSString *)emailAddressOrRawPhoneNumber displayString:(NSString * _Nullable)displayString;
+
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.h]: Provide documentation for all these class methods. Also refactor nullability as stated above.

File: CareKit/Connect/OCKContactInfo.h [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+@end
+
+NS_ASSUME_NONNULL_END
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.h]: Add extra new line at EOF.

File: CareKit/Connect/OCKContactInfo.m [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+ FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.m]: Add extra new line.

File: CareKit/Connect/OCKContactInfo.m [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+#import "OCKContactInfo.h"
+#import "OCKHelpers.h"
+#import "OCKDefines_Private.h"
+
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.m]: Add extra new line.

File: CareKit/Connect/OCKContactInfo.m [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+@implementation OCKContactInfo
+
++ (instancetype)new {
+	OCKThrowMethodUnavailableException();
+	return nil;
+}
+
+- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL {
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.m]: Remove nullability specifiers from implementation files.

File: CareKit/Connect/OCKContactInfo.m [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+			break;
+			
+		case OCKContactInfoTypeVideo:
+			defaultLabel = OCKLocalizedString(@"CONTACT_INFO_VIDEO_TITLE", nil);
+			break;
+	}
+	return [self initWithType:type displayString:displayString actionURL:actionURL label:defaultLabel];
+}
+
+- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL label:(NSString *)label {
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.m]: Remove nullability specifiers from implementation files.

File: CareKit/Connect/OCKContactInfo.m [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+			break;
+			
+		case OCKContactInfoTypeVideo:
+			defaultIcon = [UIImage imageNamed:@"video" inBundle:OCKBundle() compatibleWithTraitCollection:nil];
+			break;
+	}
+	return [self initWithType:type displayString:displayString actionURL:actionURL label:label icon:defaultIcon];
+}
+
+- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL label:(NSString *)label icon:(UIImage *)icon {
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.m]: Remove nullability specifiers from implementation files.

File: CareKit/Connect/OCKContactInfo.m [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
++ (OCKContactInfo *)facetimeVideoContactInfo:(NSString *)emailAddressOrRawPhoneNumber displayString:(NSString * _Nullable)displayString {
+	NSURL *actionURL = [[NSURL alloc] initWithString:[@"facetime://" stringByAppendingString:emailAddressOrRawPhoneNumber]];
+	return [[OCKContactInfo alloc] initWithType:OCKContactInfoTypeVideo displayString:displayString ?: emailAddressOrRawPhoneNumber actionURL:actionURL];
+}
+
++ (OCKContactInfo *)facetimeAudioContactInfo:(NSString *)emailAddressOrRawPhoneNumber displayString:(NSString * _Nullable)displayString{
+	NSURL *actionURL = [[NSURL alloc] initWithString:[@"facetime-audio://" stringByAppendingString:emailAddressOrRawPhoneNumber]];
+	return [[OCKContactInfo alloc] initWithType:OCKContactInfoTypePhone displayString:displayString ?: emailAddressOrRawPhoneNumber actionURL:actionURL];
+}
+
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContactInfo.m]: Add extra new line.

File: CareKitTests/CareKitContactTests.m [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+ FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Comment (umerkhan-apple) [inline CareKitTests/CareKitContactTests.m]: Add extra new line.

File: CareKitTests/CareKitContactTests.m [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+#import <XCTest/XCTest.h>
+#import <CareKit/CareKit.h>
+#import <CareKit/CareKit_private.h>
+#import "OCKDefines_Private.h"
+
```
Comment (umerkhan-apple) [inline CareKitTests/CareKitContactTests.m]: Add extra new line.

**Comment by umerkhan-apple** [comment]
@micahhainline Thanks for the PR. This looks great! I've completed a first-pass and provided some feedback.

Will do some device testing and provide further feedback.

**Comment by micahhainline** [comment]
@umerkhan-apple Thanks for the quick feedback! Added a new commit correcting all formatting and documentation issues that you pointed out.

**Comment by umerkhan-apple** [comment]
@micahhainline There are some merge conflicts due to a recent PR that was merged. https://github.com/carekit-apple/CareKit/pull/78

It added 3D Touch functionality to the Connect module. Could you resolve the conflicts?
Thanks! 👍

**Comment by micahhainline** [comment]
@umerkhan-apple We resolved the merge conflicts for 3D Touch. We tested on device to ensure 3D Touch options worked as before.

**Comment by umerkhan-apple** [comment]
@micahhainline This looks great! Just 2 last things.

Could you update the sample app and the test app to only use the new initializer? There are several warnings otherwise. 

And could you also update the sample app to include a video contact info item for one of the contacts?

Thanks! With those 2 changes in, this PR will be ready to be merged.

**Comment by micahhainline** [comment]
@umerkhan-apple We updated the applications to use the new init method. While doing so we realized some pain points of the long convenience function names, and adjusted them accordingly.
We noticed that some places in the test application are using old Swift syntax, mostly be fully qualifying enums when they don't have to and calling MyClass.init() instead of MyClass(). As we were moving to the new init we fixed those in the classes we touched.

**Comment by umerkhan-apple** [comment]
@micahhainline Looks great! Thanks for the quick turnaround! r=me!

**Comment by micahhainline** [comment]
@umerkhan-apple Thanks for all the support! Hopefully we'll be able to contribute more to this project in the future.

**Comment by smravca-apple** [comment]
Hi @micahhainline! Could you please provide me with a contact email so that I can send you a contribution agreement to sign? Thank you! :)

**Pull request merged**